### PR TITLE
Refactor/checkout product modal

### PR DESCRIPTION
**Pull request by futianshen**
### Refactor
`CheckoutProductMoal`
- Get `currentMember` info in `CheckoutProductModal`
- Get `product` info (isSubscription & isPhysical)  in `CheckoutProductModal`
- Remove unused props
- Fix props: renderTrigger
- Rewrite logic about syncing checkout info (invoice, shipping, payment)
   Priority: LocalStorage > Member > App > System)

`CheckoutPodcastPlanModal`
- Fix props type definition
- Remove unused props on `CheckoutPodcastPlanModal`

`useEnrolledPlanIds`
- Get `currentMemberId` in hook

### Misc
- Extract type definition from component to type file (`checkout.ts`)
- Fix type naming & definition (`AppContent.tsx`, `checkout.ts`)
- Remove `antd` Button

### CODE REVIEW

**Comment by futianshen** [comment]
[(3) 【訂閱課程】非自動續訂方案的金流調整](https://trello.com/c/Pvk0LClX/2294-3-%E3%80%90%E8%A8%82%E9%96%B1%E8%AA%B2%E7%A8%8B%E3%80%91%E9%9D%9E%E8%87%AA%E5%8B%95%E7%BA%8C%E8%A8%82%E6%96%B9%E6%A1%88%E7%9A%84%E9%87%91%E6%B5%81%E8%AA%BF%E6%95%B4)

File: src/components/checkout/CheckoutProductModal.tsx [hunk 1]
```diff
@@ -53,113 +55,117 @@ export type CheckoutProductModalProps = {
...
-    invoice: InvoiceProps
-    payment: PaymentProps
+  // checkout
+  const [productId, setProductId] = useState(defaultProductId)
+  const { target } = useSimpleProduct({ id: productId, startedAt })
+
+  // cart information
+  const memberCartInfor: {
+    shipping?: ShippingProps | null
+    invoice?: InvoiceProps | null
```
Comment (emily40830) [inline src/components/checkout/CheckoutProductModal.tsx:80]: NITS: `memberCartInfor` ---> `memberCartInfo`

File: src/components/checkout/CheckoutProductModal.tsx [hunk 2]
```diff
@@ -53,113 +55,117 @@ export type CheckoutProductModalProps = {
...
-          ? (settings['payment.perpetual.default_gateway_method'] as PaymentMethodType) || 'credit'
-          : 'credit',
-    },
+    shipping: currentMember?.shipping,
+    invoice: currentMember?.invoice,
+    payment: currentMember?.payment,
   }
 
+  const cachedCartInfor: {
+    shipping: ShippingProps | null
```
Comment (emily40830) [inline src/components/checkout/CheckoutProductModal.tsx:89]: same as above

**Review by emily40830 — APPROVED** [review]

File: src/components/checkout/CheckoutProductModal.tsx (hunk 1 above)
Comment (kkshyu) [reply src/components/checkout/CheckoutProductModal.tsx:80]: the two abbreviations are both fine, however, we use `info` more

**Comment by kkshyu** [comment]
## merge memberCardInfo and cachedCartInfo

- it is not a good idea to mix undefined and null type
- localStorage should be refetch while re-render

```ts
const memberCardInfo: {
  shipping: ShippingProps | null;
  invoice: InvoiceProps | null;
  payment: PaymentProps | null;
} | null = null;
```

```ts
const cartInfo = useCart();
type Cart
const useCart = () => {
  const [cachedCartInfo, setCachedCartInfo] = useState<{
    shipping: ShippingProps | null;
    invoice: InvoiceProps | null;
    payment: PaymentProps | null;
  }>({
    shipping: null,
    invoice: null,
    payment: null,
  });
  useEffect(() => {
    try {
      const cachedShipping = localStorage.getItem("kolable.cart.shipping");
      const cachedInvoice = localStorage.getItem("kolable.cart.invoice");
      const cachedPayment = localStorage.getItem(
        "kolable.cart.payment.perpetual"
      );
      setCachedCartInfo({
        shipping: cachedShipping && JSON.parse(cachedShipping),
        invoice: cachedInvoice && JSON.parse(cachedInvoice).value,
        payment: cachedPayment && JSON.parse(cachedPayment),
      });
    } catch {}
  }, []);
};
```

## Aggregate usePayment

`initializePayment` can be put into the effect

```ts
const [payment, setPayment] = useState<PaymentProps | null | undefined>();
const initializePayment = useCallback(
  (isProductSubscription: boolean) => {
    setPayment(
      isProductSubscription
        ? {
            gateway:
              settings["payment.subscription.default_gateway"] || "tappay",
            method: "credit",
          }
        : {
            gateway:
              settings["payment.perpetual.default_gateway"] || "spgateway",
            method:
              settings["payment.perpetual.default_gateway_method"] || "credit",
            ...memberCartInfo.payment,
            ...cachedCartInfo.payment,
          }
    );
  },
  [memberCartInfo.payment]
);
useEffect(() => {
  if (typeof target?.isSubscription === "boolean") {
    initializePayment(target.isSubscription);
  }
}, [target?.isSubscription, initializePayment]);
```

## error/loading first

```ts
if (currentMember === null || target === null || payment === undefined) {
  return renderTrigger?.({ isLoading: true });
}
```

## Aggregate refs and scroll event

in ant design, the form validation contains some methods about scrolling.
maybe use `rules` and `scrollToFirstError` will be better

FYR: https://ant.design/components/form/#scrollToFirstError-and-scrollToField-not-working-on-custom-form-control
Take care of: https://ant.design/components/form/#scrollToFirstError-and-scrollToField-not-working-on-custom-form-control

## Use app currency instead of hard coding

```ts
if (settings["tracking.fb_pixel_id"]) {
  ReactPixel.track("AddToCart", {
    value: totalPrice,
    currency: "TWD",
  });
}
if (settings["tracking.ga_id"]) {
  ReactGA.plugin.execute("ec", "addProduct", {
    id: productId,
    name: target.title || productId,
    category: productId.split("_")[0] || "Unknown",
    price: `${totalPrice}`,
    quantity: "1",
    currency: "TWD",
  });
  ReactGA.plugin.execute("ec", "setAction", "add");
  ReactGA.ga("send", "event", "UX", "click", "add to cart");
}
```


## Think about useMember

* update metadata is a kind of action of the member. it is better to put into the useMember
* maybe useCurrentMember instead of useMember


```ts
const { updateMetadata } = useCurrentMember()
```

**Pull request merged**